rules/etcd.go

```diff
@@ -85,6 +85,14 @@ func (ev3kw *etcdV3KeyWatcher) cancel() {
 	ev3kw.stopCh <- true
 }
 
+func (ev3kw *etcdV3KeyWatcher) reset() {
```

Comment:
Are you thinking that whatever resources backing up the watcher channels are being kept around because the context that was passed isn't garbage collected?

I think there is a potential for concurrency issues with this approach, because a different goroutine will call `stop` so you have one goroutine trying to interact with an instance variable that is potentially being replaced by a different goroutine.  Depending on how the compiler decides to translate this into assembly, it's even possible that the goroutine calling `stop` could get a nil pointer exception trying to call a cancel function that is in the middle of being set to a new value by another goroutine.

That being the case, I'd add a mutex and acquire it in `reset` and `stop` before manipulating the cancel function.

Reply:
Yes, from the memory profiles, I was seeing what appeared to be watcher resources being kept around.

Ah thx, a mutex is needed here.  I changed to have all cancelFunc uses call the reset to keep the mutex in one place.

Reply:
Yes, from the memory profiles, I was seeing what appeared to be watcher resources being kept around.

Ah thx, a mutex is needed here.  I changed to have all cancelFunc uses call the reset to keep the mutex in one place.